### PR TITLE
dropout and batchnorm hacks

### PR DESCRIPTION
**Pull request by smalik169**

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -315,12 +315,13 @@ def sample_batch(self, bsz, sample_sentence=SAMPLE_SENTENCE):
...
 
-        print("Source:", ''.join(map(chr, src[0].numpy())))
+        print("Source:",
+              ''.join(map(chr, src[0].numpy())).replace(chr(WILDCARD), '*'))
         yield (src.cuda(), tgt.cuda()) if self.cuda else (src, tgt)
 
 
 class UTF8CharStarFile(UTF8File):
-    def __init__(self, path, cuda, p=0.5, **kwargs):
+    def __init__(self, path, cuda, p=0.2, **kwargs):
```
Comment (alancucki) [inline data.py]: Boli, ale nie zmieniajmy parametrów w konstruktorach (stare eksperymenty mogą się posypać przy wczytywaniu). Moglibyśmy robić https://semver.org/ , wtedy trzymamy takie zmiany do wersji 2.0 :)

File: models.py [hunk 2]
```diff
@@ -83,32 +97,51 @@ def forward(self, x):
 
 class ByteCNNEncoder(nn.Module):
     def __init__(self, n, emsize, vocab_size, normalization, padding_idx=None,
-                 use_linear_layers=True, compress_channels=None):
+                 use_linear_layers=True, compress_channels=None,
+                 dropout=0.0, use_external_batch_norm=False, max_r=None):
```
Comment (alancucki) [inline models.py]: zamiast `max_r` nazwijmy to `external_batch_r`, bo się nie połapiemy za miesiąc, że `max_r` ma coś wspólnego z `use_external_batch_norm`.

File: models.py [hunk 3]
```diff
@@ -124,19 +157,33 @@ def __init__(self, n, emsize, vocab_size, normalization, padding_idx=None,
...
         x = self.prefix(x)
 
-        for _ in xrange(r-2):
-            x = self.recurrent(x)
+        assert self.max_r is None or r <= self.max_r
+        for rec_num in xrange(r-2):
+            if not self.use_external_batch_norm:
+                x = self.recurrent(x)
+            else:
+                offset = 2 * (self.n // 2) * rec_num
```
Comment (alancucki) [inline models.py]:
```
assert self.n % 2 == 0
offset = self.n * rec_num
```

File: models.py [hunk 4]
```diff
@@ -488,6 +547,11 @@ def train_on(self, batch_iterator, optimizer, logger=None):
             errs.append(err_rate)
             logger.train_log(batch, {'loss': loss.data[0], 'acc': 100. - err_rate,},
                              named_params=self.named_parameters)
+
+            if scheduler is not None:
```
Comment (alancucki) [inline models.py:555]: Trzeba jeszcze ręcznie przewinąć licznik iteracji podczas wczytywania eksperymentów. Zajmę się tym.

Comment (alancucki) [reply models.py:555]: Tj. licznik iteracji w schedulerze (poniżej).

File: models.py [hunk 5]
```diff
@@ -13,12 +13,14 @@
 from torch.autograd import Variable
 
 
-#class InstanceNorm1d(nn.BatchNorm1d):
-#    def forward(self, data):
-#        return torch.nn.functional.batch_norm(
-#            data, self.running_mean, self.running_var,
-#            self.weight, self.bias,
-#            True, self.momentum, self.eps)
+def eval_all_but_batchnorm(module): # XXX temporary solution
```
Comment (alancucki) [inline models.py:16]: Sprytnie!
